### PR TITLE
[Merged by Bors] - feat: Turán's theorem

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Clique.lean [hunk 1]
```diff
@@ -344,6 +344,34 @@ theorem cliqueFree_two : G.CliqueFree 2 ↔ G = ⊥ := by
     exact cliqueFree_bot le_rfl
 #align simple_graph.clique_free_two SimpleGraph.cliqueFree_two
 
+/-- Adding an edge increases the clique number by at most one. -/
+theorem cliqueFree_of_addEdge_cliqueFree (v w) (h : G.CliqueFree n) :
```
Comment (urkud) [inline Mathlib/Combinatorics/SimpleGraph/Clique.lean:348]:
```suggestion
protected theorem CliqueFree.addEdge (h : G.CliqueFree n) (v w) :
```

File: Mathlib/Combinatorics/SimpleGraph/Operations.lean [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+namespace SimpleGraph
+
+variable {V : Type*} [DecidableEq V] (G : SimpleGraph V) (s t : V)
+
+section ReplaceVertex
+
+/-- The graph formed by forgetting `t`'s neighbours and instead giving it those of `s`. The `s-t`
+edge is removed if present. -/
+abbrev replaceVertex : SimpleGraph V where
```
Comment (urkud) [inline Mathlib/Combinatorics/SimpleGraph/Operations.lean:33]: Why should this be an `abbrev`, not a `def`?

**Comment by fpvandoorn** [comment]
please mark as `awaiting-review` if it's ready for review

**Comment by Parcly-Taxel** [comment]
@fpvandoorn Of course it is **not** ready. The proofs are still verbose.

I'm now trying to prove `card_edgeFinset_turanGraph_add` which gives the relation between edge counts of Turán graphs **without** actually going through the 100-line proof of the absolute number of edges in the graph. But the proof terms are so complicated that I can't prove them.

**Comment by Parcly-Taxel** [comment]
@kmill I have implemented your suggestion at https://github.com/leanprover-community/mathlib4/pull/11990#discussion_r1649772433.

File: Mathlib/Combinatorics/SimpleGraph/Turan.lean [hunk 3]
```diff
@@ -280,6 +270,26 @@ theorem nonempty_iso_turanGraph : Nonempty (G ≃g turanGraph (Fintype.card V) r
 
 end IsTuranMaximal
 
-end Forward
+variable (hr : 0 < r)
+
+/-- **Turán's theorem**, reverse direction.
+
+Any graph isomorphic to `turanGraph n r` is itself Turán-maximal. -/
+theorem isTuranMaximal_of_iso (f : G ≃g turanGraph n r) : G.IsTuranMaximal r := by
```
Comment (jcommelin) [inline Mathlib/Combinatorics/SimpleGraph/Turan.lean:278]: Could you please generalize this to say that if `G1 ≃g G2` then `G1.IsTuranMaximal r` iff `G2.IsTuranMaximal r`?

Most of this proof would then turn into a proof of `isTuranMaximal_turanGraph`.

**Review by jcommelin — COMMENTED** [review]
Minor comment. Apart from that: LGTM

File: Mathlib/Combinatorics/SimpleGraph/Turan.lean (hunk 3 above)
Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/Turan.lean]: I've added the generalisation as `IsTuranMaximal.iso` – it is easier and more readable to prove the iff statement from the original theorem than to prove the iff statement directly.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**